src/passes/enumConverter.ts

```diff
@@ -24,33 +35,43 @@ export class EnumConverter extends ASTMapper {
     return val;
   }
 
-  replacementTypeString(enumDef: EnumDefinition): string {
-    return enumToIntType(enumDef).pp();
+  visitTypeName(node: TypeName, ast: AST): void {
+    this.commonVisit(node, ast);
+    const tNode = getNodeType(node, ast.compilerVersion);
+    const replacementNode = replaceEnumType(tNode);
+    if (tNode.pp() !== replacementNode.pp()) {
+      node.typeString = generateExpressionTypeString(replacementNode);
+    }
   }
 
-  visitVariableDeclaration(node: VariableDeclaration, ast: AST): void {
-    if (node.vType instanceof UserDefinedTypeName) {
-      const enumDef = node.vType.vReferencedDeclaration;
-      if (enumDef instanceof EnumDefinition) {
-        const replacementIntType = enumToIntType(enumDef);
-        const replacementIntTypeName = typeNameFromTypeNode(replacementIntType, ast);
-        ast.replaceNode(node.vType, replacementIntTypeName);
-
-        const replacementTypeString = replacementIntType.pp();
-        node.typeString = replacementTypeString;
-      }
+  visitElementaryTypeName(node: ElementaryTypeName, ast: AST): void {
+    this.commonVisit(node, ast);
+    const tNode = typeNameToTypeNode(node);
+    const replacementNode = replaceEnumType(tNode);
+    if (tNode.pp() !== replacementNode.pp()) {
+      const newTypeString = generateExpressionTypeString(replacementNode);
+      node.typeString = node.name = newTypeString;
     }
   }
 
+  visitUserDefinedTypeName(node: UserDefinedTypeName, ast: AST): void {
+    const tNode = getNodeType(node, ast.compilerVersion);
+    assert(tNode instanceof UserDefinedType, 'Expected UserDefinedType');
+    if (!(tNode.definition instanceof EnumDefinition)) return;
+    const newTypeString = generateExpressionTypeString(replaceEnumType(tNode));
+    ast.replaceNode(node, new ElementaryTypeName(node.id, node.src, newTypeString, newTypeString));
+  }
+
+  visitVariableDeclaration(node: VariableDeclaration, ast: AST): void {
+    this.commonVisit(node, ast);
+    const typeNode = replaceEnumType(getNodeType(node, ast.compilerVersion));
+    node.typeString = generateExpressionTypeString(typeNode);
+  }
+
   visitExpression(node: Expression, ast: AST): void {
     this.commonVisit(node, ast);
-    const nType = getNodeType(node, ast.compilerVersion);
-    if (nType instanceof UserDefinedType) {
-      const enumDef = nType.definition;
-      if (enumDef instanceof EnumDefinition) {
-        node.typeString = this.replacementTypeString(enumDef);
-      }
-    }
+    const typeNode = replaceEnumType(getNodeType(node, ast.compilerVersion));
+    node.typeString = generateExpressionTypeString(typeNode);
   }
 
   visitMemberAccess(node: MemberAccess, ast: AST): void {
```

Comment:
I can't seem to comment on the line in question seeing as it wasn't modified, but the createNumberLiteral here should specify a typestring based on the result of enumToInt. Should technically work either way, but doing so would make it less liable to breaking based on pass order, and should produce more efficient code

Reply:
Thanks. Implemented this.